pandas/_libs/tslib.pyx

```diff
@@ -449,26 +449,51 @@ cpdef array_to_datetime(ndarray[object] values, errors='raise',
                         dayfirst=False, yearfirst=False,
                         format=None, utc=None,
                         require_iso8601=False):
+    """
+    Converts a 1D array of date-like values to a numpy array of either:
+        1) datetime64[ns] data
+        2) datetime.datetime objects, if OutOfBoundsDatetime or TypeError
+           is encountered
+
+    Also returns a pytz.FixedOffset if an array of strings with the same
```

Comment:
In principle other tzinfos could be returned, specifically if it falls back to dateutil

Reply:
This is specifying that `array_to_datetime` function itself can return a `pytz.FixedOffset` or `None` from the C parser output. If it goes through the dateutil parser in the non-`ValueError` branch, I don't think there's a way to recover the timezone?

Reply:
When `parse_datetime_string` is called if there's a timezone it should return a tz-aware `datetime` object, so the tzinfo can be pulled off that can't it?

Reply:
Oh right that's true, good catch. Sure I will try to add some tests and functionality to hit the dateutil parser before the object branch.

Reply:
So I am using a `set` to store the parsed timezone offsets (which should be more performant that using an array in theory / I was having some trouble using an array due to duplicates), however `dateutil.tz.tzoffset`s cannot be hashed: https://github.com/dateutil/dateutil/issues/792

So instead, I am saving the `total_seconds()` of the dateutil tzoffset in the set instead and reconstructing the offsets as `pytz.FixedOffset`s